raillabel_providerkit/validation/validate_rail_side/validate_rail_side.py

```diff
@@ -0,0 +1,317 @@
+                    )
+                    right_rail: raillabel.format.Poly2d | None = _get_track_from_frame(
+                        frame, object_uid, "rightRail"
+                    )
+                    if left_rail is None or right_rail is None:
+                        continue
+
+                    swap_error: str | None = _check_rails_for_swap(left_rail, right_rail, frame_uid)
+                    if swap_error is not None:
+                        errors.append(swap_error)
+
+    return errors
+
+
+def _check_rails_for_swap(
+    left_rail: raillabel.format.Poly2d,
+    right_rail: raillabel.format.Poly2d,
+    frame_uid: str | int = "unknown",
+) -> str | None:
+    # Ensure the rails belong to the same track
+    if left_rail.object.uid != right_rail.object.uid:
+        return None
+
+    max_common_y = _find_max_common_y(left_rail, right_rail)
+    if max_common_y is None:
+        return None
+
+    left_x = _find_x_by_y(max_common_y, left_rail)
+    right_x = _find_x_by_y(max_common_y, right_rail)
+    if left_x is None or right_x is None:
+        return None
+
+    object_uid = left_rail.object.uid
+    sensor_uid = left_rail.sensor.uid if left_rail.sensor is not None else "unknown"
+
+    if left_x >= right_x:
+        return (
+            f"In sensor {sensor_uid} frame {frame_uid}, the track with"
+            f" object_uid {object_uid} has its rails swapped."
+            f" At the maximum common y={max_common_y}, the left rail has x={left_x}"
+            f" while the right rail has x={right_x}."
+        )
+
+    intersect_interval = _find_intersect_interval(left_rail, right_rail)
+    if intersect_interval is not None:
+        return (
+            f"In sensor {sensor_uid} frame {frame_uid}, the track with"
+            f" object_uid {object_uid} intersects with itself."
+            f" The left and right rail intersect in y interval {intersect_interval}."
+        )
+
+    return None
+
+
+def _count_rails_per_track_in_frame(frame: raillabel.format.Frame) -> dict[str, tuple[int, int]]:
+    # For each track, the left and right rail counts are stored as a tuple (left, right)
+    counts: dict[str, tuple[int, int]] = {}
+
+    # For each track, count the left and right rails
+    for object_uid, unfiltered_annotations in frame.object_data.items():
+        # Ensure we work only on Poly2d annotations
+        poly2ds: list[raillabel.format.Poly2d] = _filter_for_poly2ds(
+            list(unfiltered_annotations.values())
+        )
+
+        # Count left and right rails
+        left_count: int = 0
+        right_count: int = 0
+        for poly2d in poly2ds:
+            match poly2d.attributes["railSide"]:
+                case "leftRail":
+                    left_count += 1
+                case "rightRail":
+                    right_count += 1
+                case _:
+                    # NOTE: This is ignored because it is covered by validate_onthology
+                    continue
```

Comment:
It looks like match-cases were only introduced in Python 3.10 (see https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching), and since we are still supporting Python 3.8 and Python 3.9, this probably needs to be replaced by an if-else structure.